### PR TITLE
feat(streams): Stream optimizations

### PR DESCRIPTION
**Pull request by dranikpg**
Single shard XREAD becomes single hop

### CODE REVIEW

**Comment by dranikpg** [comment]
The command itself is pretty straightforward, just read the latest items from the specified streams... If it weren't for the blocking part, it could be fully a single hop command. Otherwise, we can't conditionally avoid concluding, because all shards have to give the same response, which they can't without coordinating. So... one more potential optimization is to make it full single hop if the blocking option is not specified

File: src/server/stream_family.cc [hunk 1]
```diff
@@ -2867,25 +2866,21 @@ std::optional<ReadOpts> ParseReadArgsOrReply(CmdArgList args, bool read_group,
 }
 
 // Returns the last ID of each stream in the transaction.
-OpResult<unordered_map<string_view, streamID>> FetchLastStreamIDs(Transaction* trans) {
-  vector<OpResult<vector<pair<string_view, streamID>>>> last_ids_res(shard_set->size());
-
+OpResult<absl::flat_hash_map<string, streamID>> FetchLastStreamIDs(Transaction* trans) {
+  vector<OpResult<vector<pair<string, streamID>>>> last_ids_res(shard_set->size());
```
Comment (dranikpg) [inline src/server/stream_family.cc:2870]: actually this was quite a bug 👀

Comment (romange) [reply src/server/stream_family.cc:2870]: the bug is - string_view, right?

**Review by romange — COMMENTED** [review]
looks good!

**Comment by romange** [comment]
Did you check how it's being used in bullmq? i.e. is it a blocking command? if it's being called from scripts then it's not blocking but I do not remember

File: src/server/stream_family.cc [hunk 2]
```diff
@@ -2969,77 +2964,136 @@ void XReadBlock(ReadOpts opts, ConnectionContext* cntx) {
...
-  if (!has_entries.value()) {
-    return XReadBlock(*opts, cntx);
+  if (requires_blocking)
+    return vector<RecordVec>{};
+  else
+    return {std::move(prefetched_results)};
+}
+
+// Read entries from given streams
+void XReadImpl(CmdArgList args, ReadOpts opts, ConnectionContext* cntx) {
```
Comment (romange) [inline src/server/stream_family.cc:3020]: `ReadOpts` is quite big. pls add const ref

Comment (romange) [reply src/server/stream_family.cc:3020]: here in other places

**Comment by dranikpg** [comment]
Fixed, it's used mutably. Small misdesign in general

**Review by romange — APPROVED** [review]

**Pull request merged**